core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp

```diff
@@ -83,46 +87,84 @@ class UniqueToken<Cuda, UniqueTokenScope::Global> {
+    idx = idx % size();
+#if defined(KOKKOS_ARCH_KEPLER) || defined(KOKKOS_ARCH_PASCAL) || \
+    defined(KOKKOS_ARCH_MAXWELL)
+    unsigned int mask        = __activemask();
+    unsigned int active      = __ballot_sync(mask, 1);
+    unsigned int done_active = 0;
+    bool done                = false;
+    while (active != done_active) {
+      if (!done) {
+        if (Kokkos::atomic_compare_exchange(&m_locks(idx), 0, 1) == 0) {
+          done = true;
+        } else {
+          idx += blockDim.y * blockDim.x + 1;
+          idx = idx % size();
+        }
+      }
+      done_active = __ballot_sync(mask, done ? 1 : 0);
     }
+#else
+    while (Kokkos::atomic_compare_exchange(&m_locks(idx), 0, 1) == 1) {
+      idx += blockDim.y * blockDim.x + 1;
+      idx = idx % size();
+    }
+#endif
+// Make sure that all writes in the previous lock owner are visible to me
+#ifdef KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
+    desul::atomic_thread_fence(desul::MemoryOrderAcquire(),
+                               desul::MemoryScopeDevice());
+#else
+    Kokkos::memory_fence();
+#endif
+    return idx;
+  }
 
-    return result.first;
+ public:
+  /// \brief acquire value such that 0 <= value < size()
+  KOKKOS_INLINE_FUNCTION
+  size_type acquire() const {
+    KOKKOS_IF_ON_DEVICE(return impl_acquire();)
+    KOKKOS_IF_ON_HOST(return 0;)
   }
 
   /// \brief release an acquired value
   KOKKOS_INLINE_FUNCTION
-  void release(size_type i) const noexcept {
-    Kokkos::Impl::concurrent_bitset::release(m_buffer, i);
+  void release(size_type idx) const noexcept {
+// Make sure my writes are visible to the next lock owner
+#ifdef KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
+    desul::atomic_thread_fence(desul::MemoryOrderRelease(),
+                               desul::MemoryScopeDevice());
+#else
+    Kokkos::memory_fence();
+#endif
+    (void)Kokkos::atomic_exchange(&m_locks(idx), 0);
   }
 };
 
 template <>
 class UniqueToken<Cuda, UniqueTokenScope::Instance>
     : public UniqueToken<Cuda, UniqueTokenScope::Global> {
- private:
-  Kokkos::View<uint32_t*, ::Kokkos::CudaSpace> m_buffer_view;
-
  public:
+  explicit UniqueToken() : UniqueToken<Cuda, UniqueTokenScope::Global>() {}
   explicit UniqueToken(execution_space const& arg = execution_space())
       : UniqueToken<Cuda, UniqueTokenScope::Global>(arg) {}
 
-  UniqueToken(size_type max_size, execution_space const& = execution_space())
-      : m_buffer_view(
-            "UniqueToken::m_buffer_view",
-            ::Kokkos::Impl::concurrent_bitset::buffer_bound(max_size)) {
-    m_buffer = m_buffer_view.data();
-    m_count  = max_size;
+  UniqueToken(size_type max_size) {
+    m_locks = Kokkos::View<uint32_t*, Kokkos::CudaSpace>(
+        "Kokkos::UniqueToken::m_locks", max_size);
+  }
+  UniqueToken(size_type max_size, execution_space const& exec) {
+    m_locks = Kokkos::View<uint32_t*, Kokkos::CudaSpace>(
+        Kokkos::view_alloc(exec, "Kokkos::UniqueToken::m_locks"), max_size);
```

Comment:
I slightly prefer to define and call protected constructors here to avoid allocating memory for `m_locks` twice like in https://github.com/kokkos/kokkos/pull/4748/files.

Reply:
what do you mean allocating twice?